### ISSUE TITLE
jupyter lab build fails due to NPM version.

### ISSUE THREAD

**Issue by Calvinxc1** [jupyterlab/jupyterlab#7526]
## Description

Running `jupyter lab build` ends with the following:
```
[LabBuildApp] JupyterLab 1.2.3
[LabBuildApp] Building in /opt/conda/share/jupyter/lab
[LabBuildApp] Building jupyterlab assets (build:prod:minimize)
An error occured.
RuntimeError: npm dependencies failed to install
See the log file for details:  /tmp/jupyterlab-debug-vy7qku08.log
```

The referenced log file contains the following:
```
[LabBuildApp] Building in /opt/conda/share/jupyter/lab
[LabBuildApp] Yarn configuration loaded.
[LabBuildApp] Node v6.13.1

[LabBuildApp] Building jupyterlab assets (build:prod:minimize)
[LabBuildApp] > node /opt/conda/lib/python3.7/site-packages/jupyterlab/staging/yarn.js install --non-interactive
[LabBuildApp] yarn install v1.15.2
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
error package-json@6.5.0: The engine "node" is incompatible with this module. Expected version ">=8". Got "6.13.1"
error Found incompatible module
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

[LabBuildApp] npm dependencies failed to install
[LabBuildApp] Traceback (most recent call last):

[LabBuildApp]   File "/opt/conda/lib/python3.7/site-packages/jupyterlab/debuglog.py", line 47, in debug_logging
    yield

[LabBuildApp]   File "/opt/conda/lib/python3.7/site-packages/jupyterlab/labapp.py", line 98, in start
    command=command, app_options=app_options)

[LabBuildApp]   File "/opt/conda/lib/python3.7/site-packages/jupyterlab/commands.py", line 459, in build
    command=command, clean_staging=clean_staging)

[LabBuildApp]   File "/opt/conda/lib/python3.7/site-packages/jupyterlab/commands.py", line 660, in build
    raise RuntimeError(msg)

[LabBuildApp] RuntimeError: npm dependencies failed to install

[LabBuildApp] Exiting application: JupyterLab
```

I'm running with a fresh conda environment, with JupyterHub running as a system service in that env.

I think this is related to [#5350](https://github.com/jupyterlab/jupyterlab/issues/5350).

## Reproduce

Run `jupyter lab build` in the console.

Happens even in a completely fresh env.

## Expected behavior

Expect the build to complete successfully.

## Context

- Operating System and version: Debian 9, Kernel 4.9.0-11-amd64
- Browser and version: Vivaldi 2.9.1705.41
- JupyterLab version: 1.2.3

<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
        /opt/conda/bin
        /opt/conda/condabin
        /usr/local/sbin
        /usr/local/bin
        /usr/sbin
        /usr/bin
        /sbin
        /bin

sys.path:
        /opt/conda/bin
        /opt/conda/lib/python37.zip
        /opt/conda/lib/python3.7
        /opt/conda/lib/python3.7/lib-dynload
        /opt/conda/lib/python3.7/site-packages

sys.executable:
        /opt/conda/bin/python

sys.version:
        3.7.4 (default, Aug 13 2019, 20:35:49) 
        [GCC 7.3.0]

platform.platform():
        Linux-4.9.0-11-amd64-x86_64-with-debian-9.11

which -a jupyter:
        /opt/conda/bin/jupyter
        /usr/local/bin/jupyter

pip list:
        Package                Version            
        ---------------------- -------------------
        alembic                1.3.1              
        async-generator        1.10               
        attrs                  19.3.0             
        backcall               0.1.0              
        bleach                 3.1.0              
        blinker                1.4                
        certifi                2019.9.11          
        certipy                0.1.3              
        cffi                   1.13.2             
        chardet                3.0.4              
        colorama               0.4.1              
        conda                  4.7.12             
        conda-package-handling 1.6.0              
        cryptography           2.8                
        decorator              4.4.1              
        defusedxml             0.6.0              
        entrypoints            0.3                
        gitdb2                 2.0.6              
        GitPython              3.0.5              
        idna                   2.8                
        importlib-metadata     0.23               
        ipykernel              5.1.3              
        ipython                7.9.0              
        ipython-genutils       0.2.0              
        jedi                   0.15.1             
        Jinja2                 2.10.3             
        json5                  0.8.5              
        jsonschema             3.1.1              
        jupyter-client         5.3.4              
        jupyter-core           4.6.1              
        jupyterhub             1.0.0              
        jupyterlab             1.2.3              
        jupyterlab-git         0.8.1              
        jupyterlab-server      1.0.6              
        Mako                   1.1.0              
        MarkupSafe             1.1.1              
        mistune                0.8.4              
        more-itertools         7.2.0              
        nbconvert              5.6.1              
        nbdime                 1.1.0              
        nbformat               4.4.0              
        notebook               6.0.2              
        oauthlib               3.0.1              
        pamela                 1.0.0              
        pandocfilters          1.4.2              
        parso                  0.5.1              
        pexpect                4.7.0              
        pickleshare            0.7.5              
        pip                    19.3.1             
        prometheus-client      0.7.1              
        prompt-toolkit         2.0.10             
        ptyprocess             0.6.0              
        pycosat                0.6.3              
        pycparser              2.19               
        pycurl                 7.43.0.3           
        Pygments               2.4.2              
        PyJWT                  1.7.1              
        pyOpenSSL              19.0.0             
        pyrsistent             0.15.5             
        PySocks                1.7.1              
        python-dateutil        2.8.1              
        python-editor          1.0.4              
        pyzmq                  18.1.1             
        requests               2.22.0             
        ruamel-yaml            0.15.71            
        Send2Trash             1.5.0              
        setuptools             41.6.0.post20191101
        six                    1.13.0             
        smmap2                 2.0.5              
        SQLAlchemy             1.3.11             
        terminado              0.8.3              
        testpath               0.4.4              
        tornado                6.0.3              
        tqdm                   4.38.0             
        traitlets              4.3.3              
        urllib3                1.25.7             
        wcwidth                0.1.7              
        webencodings           0.5.1              
        wheel                  0.33.6             
        zipp                   0.6.0

conda list:
        # packages in environment at /opt/conda:
        #
        # Name                    Version                   Build  Channel
        _libgcc_mutex             0.1                        main  
        alembic                   1.3.1                      py_0    conda-forge
        async_generator           1.10                       py_0    conda-forge
        attrs                     19.3.0                     py_0    conda-forge
        backcall                  0.1.0                      py_0    conda-forge
        bleach                    3.1.0                      py_0    conda-forge
        blinker                   1.4                        py_1    conda-forge
        ca-certificates           2019.9.11            hecc5488_0    conda-forge
        certifi                   2019.9.11                py37_0    conda-forge
        certipy                   0.1.3                      py_0    conda-forge
        cffi                      1.13.2           py37h8022711_0    conda-forge
        chardet                   3.0.4                 py37_1003    conda-forge
        colorama                  0.4.1                      py_0    conda-forge
        conda                     4.7.12                   py37_0    conda-forge
        conda-package-handling    1.6.0            py37h516909a_0    conda-forge
        configurable-http-proxy   4.1.0                   node6_0    conda-forge
        cryptography              2.8              py37h72c5cf5_0    conda-forge
        decorator                 4.4.1                      py_0    conda-forge
        defusedxml                0.6.0                      py_0    conda-forge
        entrypoints               0.3                   py37_1000    conda-forge
        gitdb2                    2.0.6                      py_0    conda-forge
        gitpython                 3.0.5                      py_0    conda-forge
        idna                      2.8                   py37_1000    conda-forge
        importlib_metadata        0.23                     py37_0    conda-forge
        ipykernel                 5.1.3            py37h5ca1d4c_0    conda-forge
        ipython                   7.9.0            py37h5ca1d4c_0    conda-forge
        ipython_genutils          0.2.0                      py_1    conda-forge
        jedi                      0.15.1                   py37_0    conda-forge
        jinja2                    2.10.3                     py_0    conda-forge
        json5                     0.8.5                      py_0    conda-forge
        jsonschema                3.1.1                    py37_0    conda-forge
        jupyter_client            5.3.4                    py37_0  
        jupyter_core              4.6.1                    py37_0  
        jupyterhub                1.0.0                    py37_0  
        jupyterlab                1.2.3                      py_0    conda-forge
        jupyterlab-git            0.8.2                      py_0    conda-forge
        jupyterlab_server         1.0.6                      py_0    conda-forge
        krb5                      1.16.3            h05b26f9_1001    conda-forge
        libcurl                   7.65.3               hda55be3_0    conda-forge
        libedit                   3.1.20181209         hc058e9b_0  
        libffi                    3.2.1             he1b5a44_1006    conda-forge
        libgcc                    7.2.0                h69d50b8_2    conda-forge
        libgcc-ng                 9.1.0                hdf63c60_0  
        libsodium                 1.0.17               h516909a_0    conda-forge
        libssh2                   1.8.2                h22169c7_2    conda-forge
        libstdcxx-ng              9.1.0                hdf63c60_0  
        mako                      1.1.0                      py_0    conda-forge
        markupsafe                1.1.1            py37h516909a_0    conda-forge
        mistune                   0.8.4           py37h516909a_1000    conda-forge
        more-itertools            7.2.0                      py_0    conda-forge
        nbconvert                 5.6.1                    py37_0    conda-forge
        nbdime                    1.1.0                    py37_0    conda-forge
        nbformat                  4.4.0                      py_1    conda-forge
        ncurses                   6.1               hf484d3e_1002    conda-forge
        nodejs                    6.13.1                        0    conda-forge
        notebook                  6.0.2                    py37_0  
        oauthlib                  3.0.1                      py_0    conda-forge
        openssl                   1.1.1d               h516909a_0    conda-forge
        pamela                    1.0.0                      py_0    conda-forge
        pandoc                    2.7.3                         0    conda-forge
        pandocfilters             1.4.2                      py_1    conda-forge
        parso                     0.5.1                      py_0    conda-forge
        pexpect                   4.7.0                    py37_0    conda-forge
        pickleshare               0.7.5                 py37_1000    conda-forge
        pip                       19.3.1                   py37_0    conda-forge
        prometheus_client         0.7.1                      py_0    conda-forge
        prompt_toolkit            2.0.10                     py_0    conda-forge
        ptyprocess                0.6.0                   py_1001    conda-forge
        pycosat                   0.6.3           py37h516909a_1002    conda-forge
        pycparser                 2.19                     py37_1    conda-forge
        pycurl                    7.43.0.3         py37h16ce93b_1    conda-forge
        pygments                  2.4.2                      py_0    conda-forge
        pyjwt                     1.7.1                      py_0    conda-forge
        pyopenssl                 19.0.0                   py37_0    conda-forge
        pyrsistent                0.15.5           py37h516909a_0    conda-forge
        pysocks                   1.7.1                    py37_0    conda-forge
        python                    3.7.4                h265db76_1  
        python-dateutil           2.8.1                      py_0    conda-forge
        python-editor             1.0.4                      py_0    conda-forge
        pyzmq                     18.1.1           py37h1768529_0    conda-forge
        readline                  7.0               hf8c457e_1001    conda-forge
        requests                  2.22.0                   py37_1    conda-forge
        ruamel_yaml               0.15.71         py37h14c3975_1000    conda-forge
        send2trash                1.5.0                      py_0    conda-forge
        setuptools                41.6.0                   py37_1    conda-forge
        six                       1.13.0                   py37_0    conda-forge
        smmap2                    2.0.5                      py_0    conda-forge
        sqlalchemy                1.3.11           py37h516909a_0    conda-forge
        sqlite                    3.30.1               h7b6447c_0  
        terminado                 0.8.3                    py37_0    conda-forge
        testpath                  0.4.4                      py_0    conda-forge
        tk                        8.6.9             hed695b0_1003    conda-forge
        tornado                   6.0.3            py37h516909a_0    conda-forge
        tqdm                      4.38.0                     py_0    conda-forge
        traitlets                 4.3.3                    py37_0    conda-forge
        urllib3                   1.25.7                   py37_0    conda-forge
        wcwidth                   0.1.7                      py_1    conda-forge
        webencodings              0.5.1                      py_1    conda-forge
        wheel                     0.33.6                   py37_0    conda-forge
        xz                        5.2.4             h14c3975_1001    conda-forge
        yaml                      0.1.7             h14c3975_1001    conda-forge
        zeromq                    4.3.2                he1b5a44_2    conda-forge
        zipp                      0.6.0                      py_0    conda-forge
        zlib                      1.2.11            h516909a_1006    conda-forge
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
[D 19:31:58.463 LabApp] Searching ['/mnt/storage', '/root/.jupyter', '/opt/conda/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 19:31:58.464 LabApp] Looking for jupyter_config in /etc/jupyter
[D 19:31:58.464 LabApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 19:31:58.464 LabApp] Looking for jupyter_config in /opt/conda/etc/jupyter
[D 19:31:58.464 LabApp] Looking for jupyter_config in /root/.jupyter
[D 19:31:58.464 LabApp] Looking for jupyter_config in /mnt/storage
[D 19:31:58.465 LabApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 19:31:58.465 LabApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 19:31:58.465 LabApp] Looking for jupyter_notebook_config in /opt/conda/etc/jupyter
[D 19:31:58.465 LabApp] Loaded config file: /opt/conda/etc/jupyter/jupyter_notebook_config.json
[D 19:31:58.465 LabApp] Looking for jupyter_notebook_config in /root/.jupyter
[D 19:31:58.466 LabApp] Loaded config file: /root/.jupyter/jupyter_notebook_config.json
[D 19:31:58.466 LabApp] Looking for jupyter_notebook_config in /mnt/storage
[D 19:31:58.467 LabApp] Raising open file limit: soft 1024->4096; hard 1048576->1048576
[D 19:31:58.471 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /etc/jupyter/jupyter_notebook_config.json
[D 19:31:58.471 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /usr/local/etc/jupyter/jupyter_notebook_config.json
[D 19:31:58.472 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/jupyterlab.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/nbdime.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.json
[D 19:31:58.472 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /root/.jupyter/jupyter_notebook_config.json
[D 19:31:58.749 LabApp] Using default logger
[D 19:31:58.749 LabApp] Using default logger
[D 19:31:58.749 LabApp] Using default logger
[D 19:31:58.749 LabApp] Using default logger
[D 19:31:58.749 LabApp] Using default logger
[W 19:31:58.751 LabApp] JupyterLab server extension not enabled, manually loading...
[I 19:31:59.262 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.7/site-packages/jupyterlab
[I 19:31:59.262 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[D 19:31:59.743 LabApp] Yarn configuration loaded.
[C 19:31:59.746 LabApp] Running as root is not recommended. Use --allow-root to bypass.
[D 19:31:59.746 LabApp] Exiting application: jupyter-notebook
(base) root@Karan:/mnt/storage# jupyter lab --debug --allow-root
[D 19:32:16.307 LabApp] Searching ['/mnt/storage', '/root/.jupyter', '/opt/conda/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 19:32:16.307 LabApp] Looking for jupyter_config in /etc/jupyter
[D 19:32:16.307 LabApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 19:32:16.307 LabApp] Looking for jupyter_config in /opt/conda/etc/jupyter
[D 19:32:16.308 LabApp] Looking for jupyter_config in /root/.jupyter
[D 19:32:16.308 LabApp] Looking for jupyter_config in /mnt/storage
[D 19:32:16.308 LabApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 19:32:16.308 LabApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 19:32:16.309 LabApp] Looking for jupyter_notebook_config in /opt/conda/etc/jupyter
[D 19:32:16.309 LabApp] Loaded config file: /opt/conda/etc/jupyter/jupyter_notebook_config.json
[D 19:32:16.309 LabApp] Looking for jupyter_notebook_config in /root/.jupyter
[D 19:32:16.309 LabApp] Loaded config file: /root/.jupyter/jupyter_notebook_config.json
[D 19:32:16.309 LabApp] Looking for jupyter_notebook_config in /mnt/storage
[D 19:32:16.310 LabApp] Raising open file limit: soft 1024->4096; hard 1048576->1048576
[D 19:32:16.314 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /etc/jupyter/jupyter_notebook_config.json
[D 19:32:16.315 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /usr/local/etc/jupyter/jupyter_notebook_config.json
[D 19:32:16.315 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/jupyterlab.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.d/nbdime.json
        /opt/conda/etc/jupyter/jupyter_notebook_config.json
[D 19:32:16.315 LabApp] Paths used for configuration of jupyter_notebook_config: 
        /root/.jupyter/jupyter_notebook_config.json
[D 19:32:16.585 LabApp] Using default logger
[D 19:32:16.585 LabApp] Using default logger
[D 19:32:16.586 LabApp] Using default logger
[D 19:32:16.586 LabApp] Using default logger
[D 19:32:16.586 LabApp] Using default logger
[W 19:32:16.588 LabApp] JupyterLab server extension not enabled, manually loading...
[I 19:32:17.108 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.7/site-packages/jupyterlab
[I 19:32:17.108 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[D 19:32:17.568 LabApp] Yarn configuration loaded.
[I 19:32:17.571 LabApp] Serving notebooks from local directory: /mnt/storage
[I 19:32:17.571 LabApp] The Jupyter Notebook is running at:
[I 19:32:17.571 LabApp] http://localhost:8888/?token=9bd8cfcc2e71ce278b4b1fc5b5a239d314ca684f41f9a3c7
[I 19:32:17.572 LabApp]  or http://127.0.0.1:8888/?token=9bd8cfcc2e71ce278b4b1fc5b5a239d314ca684f41f9a3c7
[I 19:32:17.572 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[W 19:32:17.576 LabApp] No web browser found: could not locate runnable browser.
[C 19:32:17.576 LabApp] 
    
    To access the notebook, open this file in a browser:
        file:///root/.local/share/jupyter/runtime/nbserver-26378-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/?token=9bd8cfcc2e71ce278b4b1fc5b5a239d314ca684f41f9a3c7
     or http://127.0.0.1:8888/?token=9bd8cfcc2e71ce278b4b1fc5b5a239d314ca684f41f9a3c7

</pre>
</details>

<details><summary>Browser Output</summary>
<pre>
N/A, haven't gotten to this point.
</pre>
</details>

**Comment by jasongrout** [comment]
Yes, please upgrade your npm like it indicates.

**Comment by Calvinxc1** [comment]
And the best approach to do that would be...? I don't think there's a more recent version of Node available in conda.

**Comment by jasongrout** [comment]
Anaconda distributes node 10, conda-forge distributes node 12: https://anaconda.org/search?q=nodejs

```
conda install nodejs
```

**Comment by Calvinxc1** [comment]
Not from what I'm seeing:

```
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /opt/conda

  added / updated specs:
    - nodejs


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.9.11          |           py37_0         154 KB
    conda-4.7.12               |           py37_0         3.0 MB
    ------------------------------------------------------------
                                           Total:         3.1 MB

The following packages will be UPDATED:

  ca-certificates    conda-forge::ca-certificates-2019.9.1~ --> pkgs/main::ca-certificates-2019.10.16-0
  openssl            conda-forge::openssl-1.1.1d-h516909a_0 --> pkgs/main::openssl-1.1.1d-h7b6447c_3

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi                                       conda-forge --> pkgs/main
  conda                                         conda-forge --> pkgs/main
  nodejs                       conda-forge::nodejs-6.13.1-0 --> pkgs/main::nodejs-6.11.2-h3db8ef7_0


Proceed ([y]/n)? 
```

**Comment by jasongrout** [comment]
What system are you on?

**Comment by Calvinxc1** [comment]
Debian 9.

**Comment by jasongrout** [comment]
Is that linux 64-bit? What if you do `conda install nodejs=12`?

**Comment by Calvinxc1** [comment]
Yep, 64-bit linux.

For some reason I'm getting a bunch of timeout errors with getting `repodata.json`. This is despite having a very healthy ping time with `repo.anaconda.com`. Hrm... I'll futz with that in the morning and let you know if it's still an issue.

**Comment by Calvinxc1** [comment]
Ok, working now, at least in regards to the timeout error. Here's what I get after running `conda install nodejs=12`:
```
(base) root@Karan:/home/jcherry# conda install nodejs=12
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - nodejs=12

Current channels:

  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.
```

**Comment by Calvinxc1** [comment]
Well, that's interesting... I ran `conda install -c conda-forge nodejs=12` and that worked, though it did take awhile:
```
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: done
```
I re-ran `jupyter lab build` afterwards and it looks like all is good.

Mind you, I had tried `conda install -c conda-forge nodejs` at one point previously, but it installed nodejs 6, not 12. Weird how I needed to specify it so exactly...

Now the new problem is even when jupyterhub is running, be at as a system service or from just running `jupyterhub` in the command line, connecting to it gives me the following error in my browser:
```
This page isn’t working
{ip to my server} didn’t send any data.
ERR_EMPTY_RESPONSE
```
This happens when connecting from a remote machine, or if I try and connect locally from the server machine itself.
